recipes/antlr4-cppruntime/all/conanfile.py

```diff
@@ -23,133 +28,118 @@ class Antlr4CppRuntimeConan(ConanFile):
-                if compiler_version < minimum_version:
-                    raise ConanInvalidConfiguration("{} requires C++17, which your compiler does not support.".format(self.name))
-            else:
-                self.output.warn("{} requires C++17. Your compiler is unknown. Assuming it supports C++17.".format(self.name))
+            if self.info.settings.compiler.cppstd:
+                check_min_cppstd(self, self._minimum_cpp_standard)
+            minimum_version = self._compilers_minimum_version.get(str(self.info.settings.compiler), False)
+            if minimum_version and Version(self.info.settings.compiler.version) < minimum_version:
+                raise ConanInvalidConfiguration(f"{self.ref} requires C++{self._minimum_cpp_standard}, which your compiler does not support.")
 
-        if is_msvc(self) and antlr_version == "4.10":
-            raise ConanInvalidConfiguration("{} Antlr4 4.10 version is broken on msvc - Use 4.10.1 or above.".format(self.name))
-
-    def build_requirements(self):
-        if self.settings.os in ("FreeBSD", "Linux"):
-            self.build_requires("pkgconf/1.7.4")
+            if is_msvc(self) and Version(self.version) == "4.10":
+                raise ConanInvalidConfiguration(f"{self.ref} is broken on msvc - Use 4.10.1 or above.")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
-
-    def _patch_sources(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
 
-    @functools.lru_cache(1)
-    def _configure_cmake(self):
-        cmake = CMake(self)
-        cmake.definitions["ANTLR4_INSTALL"] = True
-        cmake.definitions["WITH_LIBCXX"] = self.settings.compiler.get_safe("libcxx") == "libc++"
-        cmake.definitions["ANTLR_BUILD_CPP_TESTS"] = False
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["ANTLR4_INSTALL"] = True
+        tc.variables["WITH_LIBCXX"] = self.settings.compiler.get_safe("libcxx") == "libc++"
+        tc.variables["ANTLR_BUILD_CPP_TESTS"] = False
         if is_msvc(self):
-            cmake.definitions["WITH_STATIC_CRT"] = is_msvc_static_runtime(self)
-        cmake.definitions["WITH_DEMO"] = False
-        cmake.configure(build_folder=self._build_subfolder)
-        return cmake
+            tc.variables["WITH_STATIC_CRT"] = is_msvc_static_runtime(self)
+        tc.variables["WITH_DEMO"] = False
+        tc.generate()
+        tc = CMakeDeps(self)
+        tc.generate()
+        tc = VirtualBuildEnv(self)
+        tc.generate()
 
     def build(self):
-        self._patch_sources()
-        cmake = self._configure_cmake()
+        apply_conandata_patches(self)
+        cmake = CMake(self)
+        cmake.configure(build_script_folder="runtime/Cpp")
```

Comment:
Finally, I solved the linker errors when building 4.9.3 and 4.10.x with the Conan 2.0 recipe.
This "bug" in antlr's `CMakeLists.txt` was revealed after I merged [uilianries](https://github.com/uilianries)'s PR.

The CMake variable `WITH_STATIC_CRT` that controls whether `antlr` should be linked against the static MSVC runtime does not change at all. The problem is that we need to use a `CMakeLists.txt` wrapper to configure CMake, otherwise it would be too late to set that variable, and as a result `antlr` is always linked against the static MSVC runtime.

Reply:
I can build the package on Windows without any errors now. 
Let's see whether the CI build will succeed or not this time.